### PR TITLE
[TF] Implement _TensorFlowDataTypeCompatible protocol

### PR DESCRIPTION
**Pull request by njjiang**
This patch resolves [SR-8523](https://bugs.swift.org/browse/SR-8523). It implements a new protocol `_TensorFlowDataTypeCompatible`, splits the original functionality of `AccelerableByTensorFlow` into the two protocols, and makes `AccelerableByTensorFlow` refine the new protocol.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci  please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test and merge tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
The general direction of this patch is good, but with this new schema, is "AccelerableByTensorFlow" really the right thing anymore?  Shouldn't it be something like "TensorFlowNumeric"?

**Comment by rxwei** [comment]
> Shouldn't it be something like "TensorFlowNumeric"?

`Bool` isn't numeric. Maybe `TensorFlowScalar`.

**Comment by lattner** [comment]
TensorFlowScalar WFM!

File: stdlib/public/TensorFlow/TensorHandle.swift [hunk 1]
```diff
@@ -90,7 +90,8 @@ internal extension TensorHandle {
   }
 }
 
-extension TensorHandle : TensorSendableReceivable {
+extension TensorHandle : TensorSendableReceivable
+  where Scalar : AccelerableByTensorFlow {
```
Comment (mhong) [inline stdlib/public/TensorFlow/TensorHandle.swift]: will we still be able to send/receive a string tensor?

**Review by mhong — APPROVED** [review]
Very nice. Thanks you!

Sorry about the delay in review -- I was OOO last week.

File: stdlib/public/TensorFlow/TensorHandle.swift (hunk 1 above)
Comment (mhong) [reply stdlib/public/TensorFlow/TensorHandle.swift]: Sounds good.

**Review by lattner — APPROVED** [review]

**Comment by mhong** [comment]
Would you like to submit this patch, and then I can submit https://github.com/apple/swift/pull/19009 afterwards (and I'll test to make sure nothing is broken)?

**Comment by mhong** [comment]
Ah nevermind -- this PR should be orthogonal to https://github.com/apple/swift/pull/19009. Sorry about the confusion. :)

**Comment by rxwei** [comment]
@marcrasi created #20453 that gets this done. Closing this PR.

**Pull request closed without merging**